### PR TITLE
Button with icon / spinner

### PR DESCRIPTION
**Pull request by mikasdo**
https://user-images.githubusercontent.com/70565612/107225662-bc447480-6a21-11eb-8b10-8d44ef7a3424.mp4

TODO: 
 - convenience api for setting titles for different states ??
 
 *Note*: (as discussed in person) Android implementation will be in a separate PR

### CODE REVIEW

File: common/core/src/commonMain/kotlin/com/trafi/ui/component/internal/ButtonConstants.kt [hunk 1]
```diff
@@ -16,3 +17,26 @@ class ButtonConstants(theme: CurrentTheme) {
     val cornerRadius = theme.cornerRadiusScale.buttonRadius
     val defaultTruncate = true
 }
+
+class NewButtonConstants(theme: CurrentTheme) {
```
Comment (justasm) [inline common/core/src/commonMain/kotlin/com/trafi/ui/component/internal/ButtonConstants.kt:21]: Looking good! Presumably we won't be keeping around the "old" Button?

Comment (mikasdo) [reply common/core/src/commonMain/kotlin/com/trafi/ui/component/internal/ButtonConstants.kt:21]: Yes NewButton will replace Button

Comment (justasm) [reply common/core/src/commonMain/kotlin/com/trafi/ui/component/internal/ButtonConstants.kt:21]: 👌 shall we make the change with this same PR? If so, we can remove the old implementation and keep the one, new `ButtonConstants`.

File: ios/MaasCore/Sources/MaasComponents/Components/Button.swift [hunk 2]
```diff
@@ -1,75 +1,159 @@
...
+        }
+        .foregroundColor(foregroundColor)
+        .background(backgroundColor)
+        .cornerRadius(buttonConstants.cornerRadius)
+        .animation(animation)
+    }
+    
+    private var text: some View {
+        Text(input.text)
+            .lineLimit(0)
```
Comment (Ignelios) [inline ios/MaasCore/Sources/MaasComponents/Components/Button.swift:79]: Does we really need a button with more than 1 line of text ? 🤔

File: ios/MaasCore/Sources/MaasComponents/Components/Button.swift [hunk 3]
```diff
@@ -1,75 +1,159 @@
...
+    private var foregroundColor: Color {
+        input.foregroundColor ?? (isEnabled
+                                    ? buttonConstants.defaultContentColor
+                                    : buttonConstants.disabledContentColor)
+    }
+    
+    private var backgroundColor: Color {
+        input.backgroundColor ?? (isEnabled
+                                    ? buttonConstants.defaultBackgroundColor
+                                    : buttonConstants.disabledBackgroundColor)
```
Comment (Ignelios) [inline ios/MaasCore/Sources/MaasComponents/Components/Button.swift:111]: ❤️

File: ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
+struct Spinner: View {
```
Comment (Ignelios) [inline ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift:1]: Do we need some custom KMP constants for Spinner ? 🤔

File: ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+extension Animation {
+    
+    func `repeat`(while expression: Bool, autoreverses: Bool = false) -> Animation {
+        if expression {
+            return self.repeatForever(autoreverses: autoreverses)
+        } else {
+            return self
+        }
+    }
+}
```
Comment (Ignelios) [inline ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift:68]: I want to say that we should move this in to separate file and make it public... But i won't say it... 🙈

Comment (mikasdo) [reply ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift:68]: 🙈

File: ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift (hunk 4 above)
Comment (mikasdo) [reply ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift:1]: Only if they will be used in Kotlin 🤔 
But probably it makes sense to use default spinner.
@justasm do you have default spinner with Jetpack compose?

File: ios/MaasCore/Sources/MaasComponents/Components/Button.swift (hunk 2 above)
Comment (mikasdo) [reply ios/MaasCore/Sources/MaasComponents/Components/Button.swift:79]: Probably no, left this from initial `Button` implementation

File: ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift (hunk 4 above)
Comment (justasm) [reply ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift:1]: The Material implementation in Compose UI has a `CircularProgressIndicator` that looks pretty much the same and allows customizing the `color: Color` and `strokeWidth: Dp`.

Comment (mikasdo) [reply ios/MaasCore/Sources/MaasComponents/Supporting Files/CircularSpinner.swift:1]: ok then, if it's only strokeWidth, it's probably not worth defining as SpinnerConstants..

**Review by Ignelios — APPROVED** [review]
What about predefined button styles? 🤔

File: ios/MaasCore/Sources/MaasComponents/Components/Button.swift [hunk 6]
```diff
@@ -1,75 +1,160 @@
...
+    
     public struct InputType {
         public let text: String
-        public let foreground: Color?
-        public let background: Color?
+        public let icon: Image?
+        public let foregroundColor: Color?
+        public let backgroundColor: Color?
+        public let isSmall: Bool
+        public let isHugging: Bool
```
Comment (justasm) [inline ios/MaasCore/Sources/MaasComponents/Components/Button.swift:9]: 🤗 is this a term used in standard SwiftUI components? 🙂

File: ios/MaasCore/Sources/MaasComponents/Components/Button.swift [hunk 7]
```diff
@@ -80,7 +165,68 @@ public struct Button_Previews: PreviewProvider, Snapped {
...
+    @State var dFlag = false
+    
+    var title: String {
+        dFlag
+            ? isLoadingButton
+            ? ""
+            : "Button"
+            :  isLoadingButton
+            ? "Iconning..."
+            : "ICON"
```
Comment (justasm) [inline ios/MaasCore/Sources/MaasComponents/Components/Button.swift:185]: 🤯

File: ios/MaasCore/Sources/MaasComponents/Components/Button.swift [hunk 8]
```diff
@@ -80,7 +165,68 @@ public struct Button_Previews: PreviewProvider, Snapped {
...
+    var icon: Image? {
+        dFlag
+            ? Image(systemName: "command")
+            : nil
+    }
+    
+    var body: some View {
+        VStack(spacing: 20) {
+            HStack {
+                SwiftUI.Button("start") { isLoadingButton = true }
```
Comment (justasm) [inline ios/MaasCore/Sources/MaasComponents/Components/Button.swift:197]: Great to have an interactive preview 👍

**Review by justasm — APPROVED** [review]
Awesome!

File: ios/MaasCore/Sources/MaasComponents/Components/Button.swift (hunk 6 above)
Comment (mikasdo) [reply ios/MaasCore/Sources/MaasComponents/Components/Button.swift:9]: It's a term in UIKit 🙈 `contentHuggingPriority`

**Comment by mikasdo** [comment]
Let's complete this with icon and spinner, and then later do predefined button styles :)

**Pull request merged**